src/Symfony/Component/DependencyInjection/Container.php

```diff
@@ -170,24 +167,37 @@ public function set($id, $service)
             throw new InvalidArgumentException('You cannot set service "service_container".');
         }
 
-        if (isset($this->aliases[$id])) {
-            unset($this->aliases[$id]);
-        }
-
-        $this->services[$id] = $service;
-
-        if (null === $service) {
-            unset($this->services[$id]);
-        }
-
         if (isset($this->privates[$id])) {
             if (null === $service) {
                 @trigger_error(sprintf('Unsetting the "%s" private service is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0.', $id), E_USER_DEPRECATED);
                 unset($this->privates[$id]);
             } else {
                 @trigger_error(sprintf('Setting the "%s" private service is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0. A new public service will be created instead.', $id), E_USER_DEPRECATED);
             }
+        } elseif (null === $service) {
+            @trigger_error(sprintf('Unsetting the "%s" service is deprecated since Symfony 3.3 and won\'t be supported anymore in Symfony 4.0.', $id), E_USER_DEPRECATED);
+            unset($this->services[$id], $this->aliases[$id], $this->methodMap[$id]);
+
+            return;
         }
+
+        if (isset($this->aliases[$id])) {
+            @trigger_error(sprintf('Replacing the "%s" service alias is deprecated since Symfony 3.3 and will set the aliased service instead in Symfony 4.0.', $id), E_USER_DEPRECATED);
+            unset($this->aliases[$id]);
+            $this->services[$id] = $service;
+
+            return;
+        }
+
+        if (isset($this->services[$id])) {
+            @trigger_error(sprintf('Replacing the initialized "%s" service is deprecated since Symfony 3.3 and won\'t be supported anymore in Symfony 4.0.', $id), E_USER_DEPRECATED);
+        }
+
+        if (isset($this->methodMap[$id])) {
```

Comment:
or remove synthetic services from methodMap? they don't need to be their to me (neither to we have to generate any method for them), wdyt?